pkg/config/new.go

```diff
@@ -0,0 +1,238 @@
+package config
+
+import (
+	"errors"
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+	"sync"
+
+	"github.com/BurntSushi/toml"
+	"github.com/sirupsen/logrus"
+)
+
+var (
+	cachedConfigError error
+	cachedConfigMutex sync.Mutex
+	cachedConfig      *Config
+)
+
+const (
+	// FIXME: update code base and tests to use the two constants below.
+	containersConfEnv         = "CONTAINERS_CONF"
+	containersConfOverrideEnv = containersConfEnv + "_OVERRIDE"
+)
+
+// Options to use when loading a Config via New().
+type Options struct {
+	// Attempt to load the following config modules.
+	Modules []string
+
+	// Set the loaded config as the default one which can later on be
+	// accessed via Default().
+	SetDefault bool
+
+	// Additional configs to load.  An internal only field to make the
+	// behavior observable and testable in unit tests.
+	additionalConfigs []string
+}
+
+// New returns a Config as described in the containers.conf(5) man page.
+func New(options *Options) (*Config, error) {
+	if options == nil {
+		options = &Options{}
+	}
+	cachedConfigMutex.Lock()
+	defer cachedConfigMutex.Unlock()
+	return newLocked(options)
+}
+
+// Default returns the default container config.  If no default config has been
+// set yet, a new config will be loaded by New() and set as the default one.
+// All callers are expected to use the returned Config read only.  Changing
+// data may impact other call sites.
+func Default() (*Config, error) {
+	cachedConfigMutex.Lock()
+	defer cachedConfigMutex.Unlock()
+	if cachedConfig != nil || cachedConfigError != nil {
+		return cachedConfig, cachedConfigError
+	}
+	cachedConfig, cachedConfigError = newLocked(&Options{SetDefault: true})
+	return cachedConfig, cachedConfigError
+}
+
+// A helper function for New() expecting the caller to hold the
+// cachedConfigMutex.
+func newLocked(options *Options) (*Config, error) {
+	// Start with the built-in defaults
+	config, err := defaultConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	// Now, gather the system configs and merge them as needed.
+	configs, err := systemConfigs()
+	if err != nil {
+		return nil, fmt.Errorf("finding config on system: %w", err)
+	}
+	for _, path := range configs {
+		// Merge changes in later configs with the previous configs.
+		// Each config file that specified fields, will override the
+		// previous fields.
+		if err = readConfigFromFile(path, config); err != nil {
+			return nil, fmt.Errorf("reading system config %q: %w", path, err)
+		}
+		logrus.Debugf("Merged system config %q", path)
+		logrus.Tracef("%+v", config)
+	}
+
+	modules, err := options.modules()
+	if err != nil {
+		return nil, err
+	}
+
+	options.additionalConfigs = append(options.additionalConfigs, modules...)
```

Comment:
Looks like one of my comments went missing when you pushed

I don't think it makes sense to read default configs + modules for your use case.
Assuming you add modules somewhere in pkg/specgen all the defaults from the normal configs are already filled in so why should we waste time to read the same configs again just to apply the same settings.

IMO it should only read the modules in such a case.

Reply:
ok sorry I looked at the design doc (should have done that before the review) and this is a root option, I assumed somehow it was used only for create/run.

In this case this makes much more sense.

Reply:
> Assuming you add modules somewhere in pkg/specgen 

Let's take a step back and discuss when modules are being loaded.

IMO modules should only be loaded _once_ in Podman.  Somewhere in the root command we'll get modules via the `--module` CLI flag.  Then we re-load the config, set it to default and from then on, only use `Default()` throughout the stack.

Modules are extending the user and system ones, so I think the logic as is does what's intended.